locksmith/src/controllers/lockController.ts

```diff
@@ -76,18 +80,30 @@ export const stripeConnected = async (req: Request, res: Response) => {
  */
 export const lockIcon = async (req: Request, res: Response) => {
   const lockAddress = Normalizer.ethereumAddress(req.params.lockAddress)
+  const network = Number(req.params.network)
   const { original } = req.query
-  const lockIcon = await getLockIcon({
+  let icon = await getLockIcon({
     lockAddress,
     original: original === '1',
     requestUrl: Normalizer.getRequestURL(req).toString(),
   })
+  if (req.query.id) {
+    // Check if there is a custom icon for this key
+    const keyIcon = await getKeyIcon({
+      network,
+      lockAddress,
+      keyId: req.query.id.toString(),
+    })
+    if (keyIcon) {
+      icon = keyIcon
+    }
+  }
 
-  if (lockIcon.isURL) {
-    return res.redirect(lockIcon.icon)
+  if (icon.isURL) {
+    return res.redirect(icon.icon)
   } else {
-    res.setHeader('Content-Type', lockIcon.type)
-    return res.send(lockIcon.icon)
+    res.setHeader('Content-Type', icon.type!)
```

Comment:
Does the `icon.type` have to be forced because I notice sometimes in `getLockIcon` it can be `null`. Maybe some defensive coding?

```ts
if (icon.type) {
  res.setHeader('Content-Type', icon.type)
}

// or fallback to default
res.setHeader('Content-Type', icon.type ?? 'application/octet-stream')
```

Reply:
So the `icon` type is actually not great, because it can either be an actual icon, with a type, or just a URL...

Reply:
and if it is a URL we should not ave a content-type at all.

Reply:
Oh right, that's what the `isURL` is for. Food for thought later is maybe `getLockIcon` should return `IconType | string` then can use a type guard. But I see now in the meantime `type` should safely have a value if `isURL` is false.